server/src/main/java/org/elasticsearch/cluster/routing/WriteLoadPerShardSimulator.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.routing;
+
+import com.carrotsearch.hppc.ObjectDoubleHashMap;
+import com.carrotsearch.hppc.ObjectDoubleMap;
+
+import org.elasticsearch.cluster.NodeUsageStatsForThreadPools;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class WriteLoadPerShardSimulator {
+
+    private final ObjectDoubleMap<String> simulatedWriteLoadDeltas;
+    private final RoutingAllocation routingAllocation;
+    private final Map<ShardId, Double> writeLoadsPerShard;
+
+    public WriteLoadPerShardSimulator(RoutingAllocation routingAllocation) {
+        this.routingAllocation = routingAllocation;
+        this.simulatedWriteLoadDeltas = new ObjectDoubleHashMap<>();
+        writeLoadsPerShard = routingAllocation.clusterInfo().getShardWriteLoads();
+    }
+
+    public void simulateShardStarted(ShardRouting shardRouting) {
+        final Double writeLoadForShard = writeLoadsPerShard.get(shardRouting.shardId());
+        if (writeLoadForShard != null) {
```

Comment:
Could you test the case where a shard write load is 0/null? Like would be reported for a non-data stream index shard.

Reply:
I also increased the likelihood of the utilisation numbers and write loads of being zeros in a072aaf76b6b0e90fb46c991bf55576a02c348b8

We will not return a simulated utilisation of < 0.0 as well because that's nonsense

Reply:
Oh, nice change to max(X, 0.0) to avoid the negative numbers

Seems possible for the node-level and shard-level write loads not to line up exactly. E.g. the latest node-level write load reported for nodeA is 0, and it holds a shardA with peak shard write load is >0. If we then move shardA away from nodeA, we'd go negative. Shards can be moved for reasons other than write load balancing, so maybe it could happen 🧐

Reply:
Yeah I think that's something we'll have to live with, I don't think it should matter too much for the purpose of the simulation.